### PR TITLE
LLVM and Windows amd64 related improvements.

### PR DESCRIPTION
**Pull request by VictorMiasnikov**
Some LLVM and Windows amd64 related improvements:
1) LLVM9 Add GCharTriple and GCharTriple for AMD64_NT
2) Add AMD64_NT to LLVM_LIB_DIR
3) LLVM9 and AMD64_NT. Edit comment related using nmake
4) LLVM9 and AMD64_NT. Linking with psapi.lib, gdi32.lib, shell32.lib, ole32.lib, oleaut32.lib, uuid.lib, msvcprt.lib, libvcruntime.lib, libcmt.lib, oldnames.lib
5) Add AMD64_NT to LLVM_SOURCE_DIR
6) Add AMD64_NT to LLVM_BUILD_INCLUDE_DIR

### CODE REVIEW

**Comment by VictorMiasnikov** [comment]
This has been tested with LLVM 9.0.1 and MS WS 2019

File: m3-sys/llvm/llvm9/src/Main.m3 [hunk 1]
```diff
@@ -288,6 +288,11 @@ VAR
     GCharDataRep := "e-m:e-p:32:32-i32:32-i64:32:64-f80:128-n8:16:32:64-S128";
 *)
 
+(* windows x64 a-ka windows x86_64
+    GCharTriple := "x86_64-pc-windows-msvc";
+    GCharDataRep := "e-m:w-i64:64-f80:128-n8:16:32:64-S128";
+*)
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9/src/Main.m3:294]: Please call it "amd64".
Is "f80" correct?

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 2]
```diff
@@ -59,6 +61,7 @@ if not defined("LLVM_LIB_DIR")
     LLVM_LIB_DIR = PickDir("LLVM_LIB_DIR", [
         {"AMD64_LINUX"  : "/usr/local/lib", 
          "AMD64_LINUX"  : "/usr/local/llvm-3.6.1/lib", % Rodney Bates
+            "AMD64_NT"  : "/opt/llvm/lib", % VVM, de-facto it's C:\opt\llvm\lib\ or D:\opt\llvm\lib\
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:64]: `it is`

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 3]
```diff
@@ -80,7 +83,7 @@ if equal (OS_TYPE, "POSIX")
 
   exec("make", "-f" , "../src/Makefile", "-k", "LLVM_INCLUDE_DIR=" & LLVM_INCLUDE_DIR, "LLVM_BUILD_INCLUDE_DIR=" & LLVM_BUILD_INCLUDE_DIR, "BUILD_DIR=" & BUILD_DIR)
 else
-  % Windows X86
+  % Both Windows x64 and Windows X86
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:86]: `amd64`

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 4]
```diff
@@ -122,6 +125,34 @@ if equal (OS_TYPE, "POSIX")
   import_lib("dl",LIB_DIR)
   import_lib("stdc++",LIB_DIR)
   import_lib("z",LIB_DIR)
+else
+  % Both Windows x64 and Windows X86
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:129]: `amd64`

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 5]
```diff
@@ -122,6 +125,34 @@ if equal (OS_TYPE, "POSIX")
...
+  %
+  %  We should be shue that
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:131]: I don't know what `shue` is. `show`?

**Comment by mikanystrom** [comment]
I'm guessing "shue" means "sure" given the sentence construction.

**Comment by VictorMiasnikov** [comment]
Sorry - "globish" isn't my mother language.( and fafher l. too -) )I can prepare orfo checked new variant05.05.2021, 02:23, "Mika Nyström" ***@***.***>:
I'm guessing "shue" means "sure" given the sentence construction.

**Comment by VictorMiasnikov** [comment]
Can replace toWindows x64 ( a-ka amd64 or AMD64_NT)05.05.2021, 00:59, "Jay Krell" ***@***.***>:
@jaykrell commented on this pull request.



In m3-sys/llvm/llvm9bindings/src/m3makefile:
> @@ -122,6 +125,34 @@ if equal (OS_TYPE, "POSIX")
   import_lib("dl",LIB_DIR)
   import_lib("stdc++",LIB_DIR)
   import_lib("z",LIB_DIR)
+else
+  % Both Windows x64 and Windows X86

amd64

**Comment by VictorMiasnikov** [comment]
See in another answer05.05.2021, 00:59, "Jay Krell" ***@***.***>:
@jaykrell commented on this pull request.



In m3-sys/llvm/llvm9bindings/src/m3makefile:
> @@ -80,7 +83,7 @@ if equal (OS_TYPE, "POSIX")
 
   exec("make", "-f" , "../src/Makefile", "-k", "LLVM_INCLUDE_DIR=" & LLVM_INCLUDE_DIR, "LLVM_BUILD_INCLUDE_DIR=" & LLVM_BUILD_INCLUDE_DIR, "BUILD_DIR=" & BUILD_DIR)
 else
-  % Windows X86
+  % Both Windows x64 and Windows X86

amd64

**Comment by VictorMiasnikov** [comment]
It's all Ok?05.05.2021, 00:59, "Jay Krell" ***@***.***>:
@jaykrell commented on this pull request.



In m3-sys/llvm/llvm9bindings/src/m3makefile:
> @@ -59,6 +61,7 @@ if not defined("LLVM_LIB_DIR")
     LLVM_LIB_DIR = PickDir("LLVM_LIB_DIR", [
         {"AMD64_LINUX"  : "/usr/local/lib", 
          "AMD64_LINUX"  : "/usr/local/llvm-3.6.1/lib", % Rodney Bates
+            "AMD64_NT"  : "/opt/llvm/lib", % VVM, de-facto it's C:\opt\llvm\lib\ or D:\opt\llvm\lib\

it is

**Comment by VictorMiasnikov** [comment]
About f80 - I don't sure -(About amd64 - see my another e-mail message05.05.2021, 00:59, "Jay Krell" ***@***.***>:
@jaykrell commented on this pull request.



In m3-sys/llvm/llvm9/src/Main.m3:
> @@ -288,6 +288,11 @@ VAR
     GCharDataRep := "e-m:e-p:32:32-i32:32-i64:32:64-f80:128-n8:16:32:64-S128";
 *)
 
+(* windows x64 a-ka windows x86_64
+    GCharTriple := "x86_64-pc-windows-msvc";
+    GCharDataRep := "e-m:w-i64:64-f80:128-n8:16:32:64-S128";
+*)

Please call it "amd64".
Is "f80" correct?

File: m3-sys/llvm/llvm9/src/Main.m3 (hunk 1 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9/src/Main.m3:294]: 1) Fixed in https://github.com/VictorMiasnikov/cm3/commit/40c8521812e6f17f62965e65351946a400c4970b
2)

} Is "f80" correct?

I think yes, it's correct:  always "64-f80"
==
(* amd64 *)
    GCharTriple := "x86_64-pc-linux-gnu";
    GCharDataRep := "e-m:e-p:64:64-i64:64-f80:128-n8:16:32:64-S128";

(* linuxlibc6 
    GCharTriple := "i686-pc-linux-gnu";
    GCharDataRep := "e-m:e-p:32:32-i32:32-i64:32:64-f80:128-n8:16:32:64-S128";
*)

(* windows amd64
    GCharTriple := "x86_64-pc-windows-msvc";
    GCharDataRep := "e-m:w-i64:64-f80:128-n8:16:32:64-S128";
*)

(* windows 686  
    GCharTriple := "i686-pc-windows-msvc";
    GCharDataRep := "e-m:x-p:32:32-i64:64-f80:32-n8:16:32-a:0:32-S32";
*)
==

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 4 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: It's fixed. See https://github.com/VictorMiasnikov/cm3/commit/b846c7e57e251066558c684f5c6c282d8bd23b91

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 5 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: It's fixed. See https://github.com/VictorMiasnikov/cm3/commit/b846c7e57e251066558c684f5c6c282d8bd23b91

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 3 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: It's fixed. See https://github.com/VictorMiasnikov/cm3/commit/b846c7e57e251066558c684f5c6c282d8bd23b91

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 2 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: It's fixed. See https://github.com/VictorMiasnikov/cm3/commit/b846c7e57e251066558c684f5c6c282d8bd23b91

**Comment by VictorMiasnikov** [comment]
all's fixed

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 6]
```diff
@@ -122,6 +125,34 @@ if equal (OS_TYPE, "POSIX")
   import_lib("dl",LIB_DIR)
   import_lib("stdc++",LIB_DIR)
   import_lib("z",LIB_DIR)
+else
+  % Both Windows amd64 and Windows X86
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:129]: nit: Suggest remove the comment. It is just for Windows. All architectures. No need to say.

File: m3-sys/llvm/llvm9bindings/src/m3makefile [hunk 7]
```diff
@@ -80,7 +83,7 @@ if equal (OS_TYPE, "POSIX")
 
   exec("make", "-f" , "../src/Makefile", "-k", "LLVM_INCLUDE_DIR=" & LLVM_INCLUDE_DIR, "LLVM_BUILD_INCLUDE_DIR=" & LLVM_BUILD_INCLUDE_DIR, "BUILD_DIR=" & BUILD_DIR)
 else
-  % Windows X86
+  % Both Windows amd64 and Windows X86
```
Comment (jaykrell) [inline m3-sys/llvm/llvm9bindings/src/m3makefile:86]: Nit: It is just for Windows. No need to list all architectures.

**Comment by jaykrell** [comment]
Seems ok to me. I don't run this stuff. We have no CI. Just self-commit?

**Comment by jaykrell** [comment]
So Windows/amd64 with LLVM works fully? Give up on m3c? Good and bad that would be. Make a distribution?

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 7 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile:86]: } Nit: It is just for Windows. No need to list all architectures.
I think about this... Why not?

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 6 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile:129]: } Nit: It is just for Windows. No need to list all architectures.
I think about this... Why not?

**Comment by VictorMiasnikov** [comment]
} Seems ok to me
Ok =} press apply

**Comment by VictorMiasnikov** [comment]
} We have no CI.
Because don't read messages from about CI in m3devel mail list.

Joke -) or half of joke

**Comment by VictorMiasnikov** [comment]
} So Windows/amd64 with LLVM works fully?

It's run.

Hello.exe compiled and work

} Good and bad that would be.

Debug info switch is turned off

libUnicode has many compile timr error ( only with LLVM)

} Give up on m3c? 

?

} Make a distribution?

New release -- yes please make it. For "packed" on Windows.

For "LLVM on Windows" -- it's in testeing / debugging stage.
I try / will public my binary

**Comment by VictorMiasnikov** [comment]
I going to prepare
-  % Both Windows amd64 and Windows X86
+  % Windows

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 6 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile:129]: } Maybe no need to say "Windows".
Will be total delete? I am not author of first comment state...

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 7 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile:86]: } Maybe no need to say "Windows".
Will be total delete? I am not author of first comment state...

**Comment by VictorMiasnikov** [comment]
Or:
} Maybe no need to say "Windows".
Will be total delete? I am not author of first comment state...

**Comment by VictorMiasnikov** [comment]
Done:
https://github.com/VictorMiasnikov/cm3/commit/19cfaed13c43fe431116b10811bac3ec71a35df1

**Comment by VictorMiasnikov** [comment]
I will be add _needed_ comments after "press Apply button"

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 7 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: Fixed:
https://github.com/VictorMiasnikov/cm3/commit/19cfaed13c43fe431116b10811bac3ec71a35df1

File: m3-sys/llvm/llvm9bindings/src/m3makefile (hunk 6 above)
Comment (VictorMiasnikov) [reply m3-sys/llvm/llvm9bindings/src/m3makefile]: Fixed:
https://github.com/VictorMiasnikov/cm3/commit/19cfaed13c43fe431116b10811bac3ec71a35df1

**Comment by jaykrell** [comment]
Trying to reopen to I can squash and merge instead of rebase and merge..

**Comment by jaykrell** [comment]
test

**Comment by VictorMiasnikov** [comment]
} Trying to reopen to I can squash and merge instead of rebase and merge..

I press button [Restore]

} test
^^^^^^^^^ I see

**Comment by VictorMiasnikov** [comment]
I should be see by exactly this URL?
https://github.com/modula3/cm3/commits/master

--

Amd64 nt llvm #415
^^^^^^^^^^^^^^^^
I see

**Comment by VictorMiasnikov** [comment]
I see

Commits on May 6, 2021
LLVM and Windows amd64 related improvements. (#415) …

```
@jaykrell
@VictorMiasnikov
jaykrell and VictorMiasnikov committed 41 seconds ago
```

Thank!

**Pull request merged**